Add Filter.Reduce (general dimension reduction for ImageStack)

kevinyamauchi opened the pull request:
# Summary
As discussed in #1302, this component provides the ability to perform dimension reduction operations to `ImageStack` using `DataArray.reduce()`. Currently, max, mean, sum, and a user-provided  can be applied across specified dimensions. Closes #1302.

# Testing
Testing of sum, mean, and max are performed in `starfish/core/image/_filter/test/test_reduce.py`

# To Do
- [x] Add tests
- [x] Update physical coordinates

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,123 @@
+
+
+class Reduce(FilterAlgorithmBase):
+    """
```
Did we decide to put the docs in the class and not the constructor?  I can't remember. :(

kevinyamauchi replied:
This is how it's done in other filter components, but I can move it if the plan is to change it moving forward.

ttung replied:
IMO that is a mistake, and we should rectify it. :)

ttung replied:
(deleted earlier comment) I misread it.

Cool, if this is what is done elsewhere, that's fine.

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,123 @@
+        one or more Axes to project over
+    func : Union[str, Callable]
+        function to apply across the dimension(s) specified by dims.
+        If a functino is provided, it follow the form specified by
```
```suggestion
        If a function is provided, it follow the form specified by
```

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,123 @@
+    ----------
+    dims : Axes
+        one or more Axes to project over
+    func : Union[str, Callable]
```
why not accept just Callable?

dganguli replied:
i like it. this is the more 'pythonic' way in scientific computing -- will be easier for numpy/pandas users to reason about i think.

ttung replied:
@dganguli You like what?  The original or my suggestion?

kevinyamauchi replied:
I think it's nice to have the string option. I think a lot of users will find it annoying to need to import numpy.amax() and this component to perform a maximum intensity projection.

ambrosejcarr replied:
I support the string option, just for ecosystem continuity -- users are used to passing function names that are resolved by getattr, as this formalism is used a lot in `scipy.spatial.distance` among other packages.

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,123 @@
+            else:
+                raise ValueError('func should be max, mean, or sum')
+        elif callable(func):
+            warnings.warn('User-specific functions are not logged')
```
why is this true?

kevinyamauchi replied:
Ah I didn't inspect the log very closely and I thought it wasn't actually storing the callable. I now see that it does, so I can remove the warning. Thanks!

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,123 @@
+    )
+    @click.pass_context
+    def _cli(ctx, dims, func):
+        ctx.obj["component"]._cli_run(ctx, Reduce(dims, func))
```
if you make the constructor only accept callable, you can make it support a wide variety of cases by doing:

`func = getattr(np, func)`

the CLI would require someone to use `--func amax`, as that is the only one that doesn't map exactly from the expected name to name in numpy space.

dganguli replied:
oh. good point. huh.

kevinyamauchi replied:
I think this is an interesting approach and I have a couple of questions:
- Is this really accepting only a callable? As I understand it, `func` would the string of a name of a callable.
- What if the user wants to define their own function to apply across an axis or specify something that is in a different library than numpy. Would we use a try/catch on `func = getattr(np, func)`? Or is there a way to check it before trying?
- I agree that the user will expect `max` even though they would mean `amax`. Perhaps we could map them together  behind the scenes  in the constructor?
- I can replace the `func` setting logic when a string is provided with `func = getattr(np, func)` and keep the ability of users to pass their own functions as well. What do you think?

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,123 @@
+
+        # Add the reduced dims back and align with the original stack
+        reduced = reduced.expand_dims(tuple(dim.value for dim in self.dims))
+        reduced = reduced.transpose(*stack.xarray.dims)
```
I wonder if the physical coordinates are correct here.  You may want to look at #1291.

kevinyamauchi replied:
Great point. Shannon said she would merge that soon and I can fix the physical coordinates.

ttung replied:
I just landed #1353, which encompasses #1291

kevinyamauchi replied:
Awesome! Thanks @ttung and @shanaxel42 !

kevinyamauchi commented:
@dganguli , I think @ttung suggested that we keep `Filter.max_proj()` and `ImageStack.max_proj()` for a bit for API stability. I would be in favor of removing `ImageStack.max_proj()` and refactoring `Filter.max_proj()` to use `Filter.reduce()`. I think it's worth keeping `Filter.max_proj()` around because maximum intensity projection is such a common operation. Thoughts?

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,151 @@
+    def run(
+            self,
+            stack: ImageStack,
+            in_place: bool = False,
```
I would explicitly not support in_place.  It doesn't make any sense.

kevinyamauchi replied:
Sounds good. In max_proj, the `run()` method still accepts an `in_place` argument and just doesn't do anything with it. Should I do the same here? Maybe give a warning?

ttung replied:
Just remove the in_place and n_processes arguments, as neither do what they are advertised to do.

kevinyamauchi commented:
@ttung , I think I've addressed your comments. However, the test is failing during doc generation (see error below) and I'm not sure what it is expecting because it passes `flake8`
```
Warning, treated as error:
/home/travis/build/spacetx/starfish/starfish/core/image/_filter/reduce.py:docstring of starfish.image.Filter.Reduce:21:Unexpected indentation.
make[1]: *** [html] Error 2
make[1]: Leaving directory `/home/travis/build/spacetx/starfish/docs'
make: *** [docs-html] Error 2
The command "make install-dev fast" exited with 2.
```

kevinyamauchi commented:
@ttung, I think I have addressed your comments. Would you mind taking another look?

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,144 @@
+        Clip.CLIP: data above 1 are set to 1, and below 0 are set to 0
+        Clip.SCALE_BY_IMAGE: data above 1 are scaled by the maximum value, with the maximum
+        value calculated over the entire ImageStack
+        Clip.SCALE_BY_CHUNK: data above 1 are scaled by the maximum value, with the maximum
```
group_by is not provided...?

kevinyamauchi replied:
Ah good point. I will remove this option.

ttung commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,144 @@
+        max: maximum intensity projection (applies numpy.amax)
+        mean: take the mean across the dim(s) (applies numpy.mean)
+        sum: sum across the dim(s) (applies numpy.sum)
+    clip_method : Union[str, Clip]
```
should just be
 
```suggestion
    clip_method : Clip
```

kevinyamauchi replied:
Thanks! Fixed.

ambrosejcarr commented on starfish/core/image/_filter/reduce.py:
```diff
@@ -0,0 +1,144 @@
+        matches the form specified above
+        (i.e., function is resolved: func = getattr(np, func)).
+        Some common examples below:
+        max: maximum intensity projection (applies numpy.amax)
```
@kevinyamauchi `max` -> `np.amax` is not consistent with the stated functionality of the documentation using `getattr`, and in the code, it's special cased. Can you resolve the logical discontinuity in the documentation?

kevinyamauchi replied:
Ah good point. Fixed.
